### DISCUSSION TITLE
How to determine the --batch_size when training from scratch?

### DISCUSSION THREAD

**Discussion by lishuai-97** [mlfoundations/open_clip#848] [Q&A] [answered]
How do you determine the `--batch_size`  of  ViT-B-32 model for Laion-2B? For example, this [run](https://wandb.ai/rom1504/open-clip/runs/2xpw65p8/overview?workspace=) uses --batch-size = 352 when trian on 8 NVIDIA A100-SXM4-40GB GPUs.

However, when I was training from scratch on a ViT-B-32 model with 8 x 4090 GPUs using the CC3M dataset (download from [here](https://huggingface.co/datasets/pixparse/cc3m-wds) by using the `datasets` package) with the script below, I found that a `batch-size` of 352 only occupied 7-8GB of VRAM on each GPU. Furthermore, the `batch-size` could be set up to 2048, which roughly consumed about 20GB of VRAM per GPU. However, in the provided wandb settings, you used  A100 GPUs with 40GB VRAM, but your `batch-size` was only 352, which is quite odd. Apart from the difference in datasets, all my other parameters are the same as your wandb hyperparameters. So, I would like to know, is there any unmentioned requirement for the `batch-size` per GPU when training from scratch? Or did I miss any other parameter settings? 

```bash
# Single-Node
torchrun --nproc_per_node 8 -m training.main \
    --save-frequency 10 \
    --train-data 'data/cc3m/cc3m-train-{0000..0575}.tar::data/cc3m/cc3m-validation-{0000..0015}.tar' \
    --train-num-samples 135646078 \
    --dataset-type webdataset \
    --precision amp_bf16 \
    --warmup 5000 \
    --batch-size 352 \
    --epochs 150 \
    --dataset-resampled \
    --lr 2e-3 \
    --beta1 0.9 \
    --beta2 0.99 \
    --lr-scheduler cosine \
    --wd 0.2 \
    --force-patch-dropout 0.5 \
    --report-to tensorboard \
    --workers 4 \
    --model ViT-B-32 \
    --name "ViT-B-32-Vanilla-1" \
    --log-every-n-steps 1 \
    --seed 0 \
    --ddp-static-graph \
    --local-loss \
    --gather-with-grad \
    --grad-checkpointing \
```

 @mitchellnw @rom1504 @rwightman

**Comment by rom1504** [comment, marked as answer]
Usually a total batch size of at least 32k is needed to get good results
with clip. Increasing up to 64k sometimes helps

**Reply by lishuai-97** [reply]
Thank you for your suggestion, but I think I did not clearly express my confusion. In your wandb description, you used an A100 GPU with 40GB of VRAM and set the `batch-size` to `352`, whereas I used a 4090 GPU with 24GB of VRAM and was able to set the `batch-size` to `2048`. I am wondering if this is reasonable.

**Reply by lishuai-97** [reply]
Thank you, I have resolved the confusion. After using gradient accumulation to increase the total batch size, the `batch-size` parameter became consistent with the scale of your experiments.
